tools/README.md

```diff
@@ -0,0 +1,193 @@
+- `mmtk:bucket_opened(id: int)`: a work bucket opened. The first argument is the numerical representation of `enum WorkBucketStage`.
+- `mmtk:work_poll()`: a work packet is to be polled.
+- `mmtk:work(type_name: char *, type_name_len: int)`: a work packet was just executed. The first argument is points to the string of the Rust type name of the work packet, and the second argument is the length of the string.
+- `mmtk:alloc_slow_once_start()`: the allocation slow path starts.
+- `mmtk:alloc_slow_once_end()`: the allocation slow path ends.
+
+## Running tracing tools
+The tracing tools are to be invoked by a wrapper script `run.py`.
+```
+usage: run.py [-h] [-b BPFTRACE] -m MMTK [-H] [-p] [-f {text,json}] tool
+
+positional arguments:
+  tool                  Name of the bpftrace tool
+
+optional arguments:
+  -h, --help            show this help message and exit
+  -b BPFTRACE, --bpftrace BPFTRACE
+                        Path of the bpftrace executable
+  -m MMTK, --mmtk MMTK  Path of the MMTk binary
+  -H, --harness         Only collect data for the timing iteration (harness_begin/harness_end)
+  -p, --print-script    Print the content of the bpftrace script
+  -f {text,json}, --format {text,json}
+                        bpftrace output format
+```
+
+- `-b`: the path to the `bpftrace` executable. By default, it uses `bpftrace` executable in your `PATH`. We strongly recommend you use the latest statically complied `bpftrace` from [upstream](https://github.com/iovisor/bpftrace/releases). You need to be able to have sudo permission for whichever `bpftrace` you want to use.
+- `-m`: the path to a MMTk binary that contains the tracepoints. This depends on the binding you use. For the OpenJDK binding, it should be `jdk/lib/server/libmmtk_openjdk.so` under your build folder. To check whether the binary contains tracepoints, you can use `readelf -n`. You should see a bunch of `stapsdt` notes with `mmtk` as the provider.
+- `-H`: pass this flag is you want to only measure the timing iteration of a benchmark. By default, the tracing tools will measure the entire execution.
+- `-p`: print the entire tracing script before execution. This is mainly for debugging use.
+- `-f`: change the bpftrace output format. By default, it uses human-readable plain text output (`text`). You can set this to `json` for easy parsing.
+
+Please run the tracing tools **before** running the workload. If you use `-H`, the tracing tools will automatically end with `harness_end` is called. Otherwise, you will need to terminate the tools manually with `Ctrl-C`. These tools also have a timeout of 1200 seconds so not to stall unattended benchmark execution. 
+
+## Tracing tools
+### Measuring the time spend in allocation slow path (`alloc_slow`)
+This tool measures the distribution of the allocation slow path time.
+The time unit is ns / 400, so that we use the histogram bins with higher fidelity better.
```

Comment:
"The time unit is ns / 400" means "1 in the output means (1/400) ns, i.e. 0.0025 nanoseconds, i.e. 2.5 picoseconds".  That's probably not what the output really represents.  The next paragraph says it is 3.2 us to 6.4 us, i.e  8 * 400 ns to 16 * 400 ns.  So the unit should be "400 ns".

```suggestion
The time unit is 400 ns, so that we use the histogram bins with higher fidelity better.
```

Reply:
Fixed